enhancements/single-node-production-deployment-approach.md

```diff
@@ -0,0 +1,519 @@
+the operators stabilize. Users should wait for this period of
+instability before installing their workloads.
+
+The single-node configuration of OpenShift will be sufficiently
+generic to cater to a variety of edge computing use cases. As such,
+OpenShift's usual cluster configuration mechanisms will be favored
+where there is a likelihood that there will be edge computing use case
+with differing requirements.  For example, there is no expectation
+that single-node deployments will use the real-time kernel by
+default -- this will continue to be a `MachineConfig` choice as per
+[enhancements/support-for-realtime-kernel](https://github.com/openshift/enhancements/blob/master/enhancements/support-for-realtime-kernel.md). *See open questions*
+
+### User Stories
+
+#### As a user, I can deploy OpenShift in a supported single-node configuration
+
+A user will be able to run the OpenShift installer to create a single-node
+deployment, with some limitations (see non-goals above). The user
+will not require special support exceptions to receive technical assistance
+for the features supported by the configuration.
+
+### Implementation Details/Notes/Constraints
+
+Some OpenShift components require a minimum of 2 or 3 nodes. The new
+cluster capabilities API will be used to configure these components as
+appropriate for a single node.
+
+#### installer
+
+We will add logic to the installer to set the new
+`highAvailabilityMode` status field of the `infrastructure` API
+resource. The will value be set to `Full` by default and `None` when
+the number of replicas for the control plane less than `3`. See the [cluster
+high-availability mode
+API](https://github.com/openshift/enhancements/pull/555) enhancement
+for details.
+
+#### machine-config-operator
+
+The machine-config-operator includes a check in code that the number
+of control plane nodes is 3. This code needs to be updated to look at
+the new high-availability mode capability API to determine the number
+of expected nodes to allow the operator to complete its work, for
+example to enable the realtime kernel on a single node with the
+appropriate performance profile settings.
+
+The machine-config-operator tries to drain a node before applying
+local changes and rebooting. In a true single-node deployment, there
+is nowhere for drained pods to be rescheduled. In a deployment with a
+single-node control plane and additional workers, some nodes could
+potentially be drained, but there is no way for the MCO to know if the
+workers are fully utilized. When the high-availability mode is `None`
+the step needs to be skipped. Ideally we will be able to rely on the
+[graceful shutdown support in
+kubelet](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown)
+to stop all workloads safely as part of the reboot. That feature is
+alpha in kubernetes 1.20 and disabled by default, so we will need to
+add a feature gate to enable it.
+
```

Comment:
```suggestion

Any operator that generates a MachineConfig and templates in the machine-config-operator must high-availability mode agnostic. 

```

Reply:
I missed this comment earlier. Maybe we want to fold it into #587?